---
title: "Imbalanced integration guidelines tutorial - R"
output: rmarkdown::html_vignette
date: '2023-04-24'
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy = TRUE
)
```

This tutorial is based on the [Imbalanced integration guidelines](https://www.biorxiv.org/content/10.1101/2022.10.06.511156v3.full) in the Iniquitate manuscript. This example explores integration in the case of a severely imbalanced dataset - namely dataset number 7 from the [Tran et al. scRNA-seq integration benchmarking paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9#availability-of-data-and-materials). Current scRNA-seq integration techniques perform very poorly on this dataset, as illustrated by the results from Tran et al. when benchmarking on this data. As such, we shouldn't expect a strong integration performance when it comes to conserving biological heterogeneity, but we'll try our best to tune this as we pretend we don't know the ground-truth labels. 

We'll begin by illustrating the properties of this dataset comprised of two batches of mouse retinal cells. 

## Dataset exploration 

Start by downloading the data - both batches:
```{r, results="hide"}
library(googledrive)

# Create data directory
if (!dir.exists("data")){
  dir.create("data")  
}

# Install files if not already present
if (!(file.exists("data/tran_et_al_batch_1.h5ad"))) {    
  # Remove authorization for Google Drive
  drive_deauth()
  drive_user()
  
  # Download the files
  file_id_1 <- as_id("1LryRxnVGq2Ny4awDxyiS84oou_zkvQki")
  file_id_2 <- as_id("1JMhR3cGGj4pueSLiFCPYl19jZDUKmznR")
  file_1 <- drive_get(file_id_1)
  file_2 <- drive_get(file_id_2)
  googledrive::drive_download(
    file_1, 
    path = "data/tran_et_al_batch_1.h5ad", 
    overwrite=TRUE
  )
  googledrive::drive_download(
    file_2,
    path = "data/tran_et_al_batch_2.h5ad",
    overwrite=TRUE
  )
}
```

Read in both files using the SeuratDisk library
```{r, results="hide"}
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
remotes::install_github("mojaveazure/seurat-disk")
library(SeuratDisk)

# Convert the tran et al files to h5seurat format
Convert(
  "data/tran_et_al_batch_1.h5ad", 
  dest = "h5seurat", 
  overwrite = TRUE
)
Convert(
  "data/tran_et_al_batch_2.h5ad",
  dest = "h5seurat",
  overwrite = TRUE
)

# Load h5ad files for both batches
retina_1 <- SeuratDisk::LoadH5Seurat("data/tran_et_al_batch_1.h5seurat")
retina_2 <- SeuratDisk::LoadH5Seurat("data/tran_et_al_batch_2.h5seurat")
```

Examine the cell-type distribution for each batch
```{r, fig.width=14, fig.height=6, fig.align= "center"}
library(ggplot2)
library(ggthemes)

# Grab the metadata
meta_1 <- retina_1@meta.data
meta_2 <- retina_2@meta.data

# Print and plot the cell-type distribution for each batch
print(table(meta_1$celltype))
ggplot(data = meta_1, aes(celltype,fill=as.factor(celltype))) + 
  geom_histogram(stat="count") +
  labs(
    fill = "Cell-type",
    x = "Cell-type",
    y = "Count in batch 1"
  ) +
  theme_few() +
  theme(axis.title.x = element_text(size = 16)) +
  theme(axis.title.y = element_text(size = 16)) +
  theme(strip.text.x = element_text(size = 16)) +
  theme(plot.title = element_text(size = 14)) +
  theme(axis.text.x = element_text(size = 16)) +
  theme(axis.text.y = element_text(size = 16)) +
  theme(legend.title = element_text(size = 16)) +
  theme(legend.text = element_text(size = 16))

print(table(meta_2$celltype))
ggplot(data = meta_2, aes(celltype,fill=as.factor(celltype))) + 
  geom_histogram(stat="count") +
  labs(
    fill = "Cell-type",
    x = "Cell-type",
    y = "Count in batch 2"
  ) +
  theme_few() +
  theme(axis.title.x = element_text(size = 16)) +
  theme(axis.title.y = element_text(size = 16)) +
  theme(strip.text.x = element_text(size = 16)) +
  theme(plot.title = element_text(size = 14)) +
  theme(axis.text.x = element_text(size = 16)) +
  theme(axis.text.y = element_text(size = 16)) +
  theme(legend.title = element_text(size = 16)) +
  theme(legend.text = element_text(size = 16))
```
As we can see, there is an extreme imbalance in terms of the cell-types present in both batches, and this is also a special case of imbalance where quite a few cell-types (astrocytes, fibroblasts, ganglion, horizontal, microglia, pericytes, vascular endothelium) are only present in batch 2 but not batch one.

Now that we've seen the imbalance present in this dataset, we're going to take the role of a single-cell analysis practitioner that **does not know the underlying cell-type distribution of this dataset**. I.e. we don't know what cell-types are present and our goal is to learn a joint representation of these two batches of retinal cells that we've obtained from an experiment. 

We'll start from the top of the guidelines and work our way down.

## 1) Pre-integration stage

Let's make our way down the flowchart, starting from the pre-integration stage.

To perform our analysis workflow, we'll use the `SingleCellExperiment` R package and associated bioconductor packages. Lets first convert our Seurat class objects for the two batches to SingleCellExperiment (sce) objects.

```{r, results="hide"}
library(Seurat)

retina_1_sce <- as.SingleCellExperiment(retina_1)
retina_2_sce <- as.SingleCellExperiment(retina_2)
```

### Unsupervised clustering within each batch 

```{r}
library(Seurat)

# Preprocess the data using Seurat - Normalize and scale  
retina_1 <- NormalizeData(
  retina_1, normalization.method = "LogNormalize", scale.factor = 10000
)
retina_1 <- FindVariableFeatures(retina_1, nfeatures = 2000)
all.genes <- rownames(retina_1)
retina_1 <- ScaleData(retina_1, features = all.genes)

# Run PCA, create neighborhood graph and get the clusters using the Leiden
# algorithm 
retina_1 <- RunPCA(retina_1, features = VariableFeatures(retina_1))
retina_1 <- FindNeighbors(retina_1, dims = 1:10)
retina_1 <- FindClusters(retina_1, resolution = 0.5, algorithm = "leiden")

# Get the UMAP reduction of the data
retina_1 <- RunUMAP(retina_1, umap.method = "umap-learn", dims = 1:10)

# Do the same for retina 2 
retina_2 <- NormalizeData(
  retina_2, normalization.method = "LogNormalize", scale.factor = 10000
)
retina_2 <- FindVariableFeatures(retina_2, nfeatures = 2000)
all.genes <- row.names(retina_2)
retina_2 <- ScaleData(retina_2, features = all.genes)
retina_2 <- RunPCA(retina_2, features = VariableFeatures(retina_2))
retina_2 <- FindNeighbors(retina_2, dims = 1:10)
retina_2 <- FindClusters(retina_2, resolution = 0.5, algorithm = "leiden")
retina_2 <- RunUMAP(retina_2, umap.method = "umap-learn", dims = 1:10)
```

# Plot the UMAPs of the resulting clustering
```{r}
DimPlot(retina_1, reduction = "umap")
DimPlot(retina_2, reduction = "umap")
```

# Examine the number of clusters and proportions 
```{r}
library(ggplot2)

# Update metadata 
meta_1 <- retina_1@meta.data
meta_2 <- retina_2@meta.data

# Plot the number of clusters and proportions for both datasets
ggplot(meta_1, aes(x=seurat_clusters)) + geom_bar()
ggplot(meta_2, aes(x=seurat_clusters)) + geom_bar()
  
```

# There are differences, integrate with Seurat first
```{r}
library(Seurat)

retina_1@meta.data$batch <- "Batch 1"
retina_2@meta.data$batch <- "Batch 2"

retina_list <- c(retina_1, retina_2)
features <- SelectIntegrationFeatures(object.list = retina_list)
anchors <- FindIntegrationAnchors(
  object.list = retina_list, anchor.features = features
)
retina_integrated <- IntegrateData(anchorset=anchors)


retina_integrated <- ScaleData(retina_integrated)
retina_integrated <- RunPCA(retina_integrated)
retina_integrated <- FindNeighbors(retina_integrated, reduction="pca", dims=1:15)
retina_integrated <- FindClusters(retina_integrated, resolution=0.5, algorithm="leiden")
retina_integrated <- RunUMAP(retina_integrated, reduction="pca", dims=1:15)

```


# Plot the results
```{r}
DimPlot(retina_integrated, reduction="umap", group.by="batch")
DimPlot(retina_integrated, reduction="umap", group.by="seurat_clusters")
```

# Measure the degree of batch mixing 
```{r}
library(bluster)
batch_labels <- retina_integrated@meta.data$batch
integrated_cluster_labels <- retina_integrated@meta.data$seurat_clusters
a <- 1 - pairwiseRand(batch_labels, integrated_cluster_labels)
a
```

# Strong batch mixing - check diagnostics for cell-type heterogeneity conservation

```{r}
library(pheatmap)
library(bluster)

cluster_labels_b1_before <- retina_1@meta.data$seurat_clusters
cluster_labels_b2_before <- retina_2@meta.data$seurat_clusters

cluster_labels_b1_after <- retina_integrated@meta.data$seurat_clusters[
  retina_integrated@meta.data$batch == "Batch 1"
]
cluster_labels_b2_after <- retina_integrated@meta.data$seurat_clusters[
  retina_integrated@meta.data$batch == "Batch 2"
]

tab_b1 <- nestedClusters(
  ref=paste("Before", cluster_labels_b1_before), 
  alt=paste("After", cluster_labels_b1_after)
)
tab_b2 <- nestedClusters(
  ref=paste("Before", cluster_labels_b2_before),
  alt=paste("After", cluster_labels_b2_after)
)

heat_b1 <- pheatmap(
  tab_b1$proportions,
  cluster_row=FALSE,
  cluster_col=FALSE,
  main="Retina Batch 1 Comparison",
  silent=TRUE
)
heat_b2 <- pheatmap(
  tab_b2$proportions,
  cluster_row=FALSE,
  cluster_col=FALSE,
  main="Retina Batch 2 Comparison",
  silent=TRUE
)

gridExtra::grid.arrange(heat_b1, heat_b2)
```